### PR TITLE
setup unit tests for binding intermediate representation

### PR DESCRIPTION
**Pull request by jkurdek**

### CODE REVIEW

File: src/Controls/src/BindingSourceGen/BindingSourceGenerator.cs [hunk 1]
```diff
@@ -15,55 +15,186 @@ namespace Microsoft.Maui.Controls.BindingSourceGen;
...
+			types.Add("identifier");
+			var member = identifier.Identifier.Text;
+			var typeInfo = context.SemanticModel.GetTypeInfo(identifier).Type;
+			if (typeInfo == null)
+			{
+				return;
+			}; // TODO
+			var isNullable = IsNullable(typeInfo);
+			parts.Add(new PathPart(member, isNullable));
+			return;
```
Comment (simonrozsival) [inline src/Controls/src/BindingSourceGen/BindingSourceGenerator.cs:143]: Please change this to `if ... else if ... else if`. The `return` at the end of each `if` seems error prone and easy to unintentionally remove sometime later.

File: src/Controls/src/BindingSourceGen/BindingSourceGenerator.cs [hunk 2]
```diff
@@ -15,55 +15,186 @@ namespace Microsoft.Maui.Controls.BindingSourceGen;
...
+		var pathasString = string.Join("<>", parts.Select(p => p.MemberName));
+		var typesAsString = string.Join("<>", types);
+
+
+		return new Binding(
+			Id: ++_idCounter,
+			Location: sourceCodeLocation,
+			SourceType: new TypeName(inputTypeGlobalPath, IsNullable(inputType), inputTypeIsGenericParameter),
+			PropertyType: new TypeName(outputTypeGlobalPath, IsNullable(outputType), outputTypeIsGenericParameter),
+			Path: [.. parts],
```
Comment (simonrozsival) [inline src/Controls/src/BindingSourceGen/BindingSourceGenerator.cs:121]: I'd prefer using `parts.ToArray()` here to make it clearer that this copies the list into an array.

File: src/Controls/src/BindingSourceGen/BindingSourceGenerator.cs [hunk 3]
```diff
@@ -15,55 +15,186 @@ namespace Microsoft.Maui.Controls.BindingSourceGen;
...
+		var outputTypeIsGenericParameter = outputType.Kind == SymbolKind.TypeParameter;
+
+		var parts = new List<PathPart>();
+		var types = new List<string>();
+		ParsePath(lambda.Body, context, parts, types);
+
+
+		// Convert to string
+		var pathasString = string.Join("<>", parts.Select(p => p.MemberName));
+		var typesAsString = string.Join("<>", types);
```
Comment (simonrozsival) [inline src/Controls/src/BindingSourceGen/BindingSourceGenerator.cs:113]: Will we need this later? Are these lines of code used for debugging purposes?

File: src/Controls/src/BindingSourceGen/BindingSourceGenerator.cs [hunk 4]
```diff
@@ -15,55 +15,186 @@ namespace Microsoft.Maui.Controls.BindingSourceGen;
 [Generator(LanguageNames.CSharp)]
 public class BindingSourceGenerator : IIncrementalGenerator
 {
-	public void Initialize(IncrementalGeneratorInitializationContext initializationContext)
+	// TODO:
+	// Diagnostics
+	// Edge cases
+	// Optimizations
+	static int _idCounter = 0;
```
Comment (simonrozsival) [inline src/Controls/src/BindingSourceGen/BindingSourceGenerator.cs:22]: I don't think we need a custom counter (see https://github.com/dotnet/aspnetcore/blob/f9c4b98947d606088b29a305bdf1ea641576d879/src/Http/Http.Extensions/gen/RequestDelegateGenerator.cs#L67C98-L67C119)

File: src/Controls/src/BindingSourceGen/BindingSourceGenerator.cs [hunk 5]
```diff
@@ -15,55 +15,186 @@ namespace Microsoft.Maui.Controls.BindingSourceGen;
...
+		if (expressionSyntax is MemberAccessExpressionSyntax memberAccess)
+		{
+			types.Add("memberAccess");
+			var member = memberAccess.Name.Identifier.Text;
+			var typeInfo = context.SemanticModel.GetTypeInfo(memberAccess.Expression).Type;
+			if (typeInfo == null)
+			{
+				return;
+			};
+			ParsePath(memberAccess.Expression, context, parts, types); //TODO: Nullable
```
Comment (simonrozsival) [inline src/Controls/src/BindingSourceGen/BindingSourceGenerator.cs:155]: I wonder if we should use recursion. We could have just a `while (expressionSyntax is not null)` loop and just keep moving along the expression. We should keep performance in mind from the start :)

File: src/Controls/src/BindingSourceGen/BindingSourceGenerator.cs [hunk 6]
```diff
@@ -15,55 +15,186 @@ namespace Microsoft.Maui.Controls.BindingSourceGen;
...
+			Location: sourceCodeLocation,
+			SourceType: new TypeName(inputTypeGlobalPath, IsNullable(inputType), inputTypeIsGenericParameter),
+			PropertyType: new TypeName(outputTypeGlobalPath, IsNullable(outputType), outputTypeIsGenericParameter),
+			Path: [.. parts],
+			GenerateSetter: true //TODO: Implement
+		);
 	}
 
-	private bool IsSetBindingMethod(SyntaxNode node)
+	static void ParsePath(CSharpSyntaxNode? expressionSyntax, GeneratorSyntaxContext context, List<PathPart> parts, List<string> types)
```
Comment (simonrozsival) [inline src/Controls/src/BindingSourceGen/BindingSourceGenerator.cs:126]: What is the `types` list for? Is it just for debugging purposes?

Comment (jkurdek) [reply src/Controls/src/BindingSourceGen/BindingSourceGenerator.cs:126]: yup, debug purpose only, will be removed

File: src/Controls/src/BindingSourceGen/BindingSourceGenerator.cs (hunk 5 above)
Comment (jkurdek) [reply src/Controls/src/BindingSourceGen/BindingSourceGenerator.cs:155]: I found out that the root of the expression is not necessericaly the first element. So if we have e.g. `foo?.property` the first thing we find is actually conditional access. Then in order to get `foo` we need to parse `conditionalAccess.Expression`. If this is the only edge case we can properly use simple while, if not we will have to choose between BFS or recursion. I will take a look into that.

File: src/Controls/src/BindingSourceGen/BindingSourceGenerator.cs (hunk 3 above)
Comment (jkurdek) [reply src/Controls/src/BindingSourceGen/BindingSourceGenerator.cs:113]: It is only used for debug

File: src/Controls/src/BindingSourceGen/BindingSourceGenerator.cs (hunk 5 above)
Comment (simonrozsival) [reply src/Controls/src/BindingSourceGen/BindingSourceGenerator.cs:155]: Yeah, I would expect that.

BTW the conditional access in the expression what we should use as indication of "is nullable". But getting the nullable annotations right will be definitely one of the "remaining 20%" problems.

File: src/Controls/src/BindingSourceGen/BindingSourceGenerator.cs (hunk 4 above)
Comment (jkurdek) [reply src/Controls/src/BindingSourceGen/BindingSourceGenerator.cs:22]: Index there comes from custom functions defined in asp net core:
https://github.com/dotnet/aspnetcore/blob/main/src/Shared/RoslynUtils/IncrementalValuesProviderExtensions.cs

Let's keep counter for now, and figure a better way later.

File: src/Controls/src/BindingSourceGen/BindingSourceGenerator.cs (hunk 5 above)
Comment (simonrozsival) [reply src/Controls/src/BindingSourceGen/BindingSourceGenerator.cs:155]: OK, let's keep it for now and we can refactor it later.

**Review by simonrozsival — APPROVED** [review]

**Pull request merged**